### DISCUSSION TITLE
Update FGTrimAnalysis wrt latest JSBSim Release

### DISCUSSION THREAD

**Discussion by rishi1906** [JSBSim-Team/jsbsim#988] [General]
Greeting @agodemar  ,

I am new to JSBSim, trying to use the trim solution provided by Dr Agostino De Marco as mentioned in his paper "A General Solution to the Aircraft Trim Problem". When I add the header and source files of FGTrimAnalysis and FGTrimAnalysisControl to CMakeLists.txt under /src/initialization folder to build, it throws several errors of incompatibility (missing/mismatched function). 

Among several errors, one error looks like this 

../../jsbsim-code/src/initialization/FGTrimAnalysis.cpp:335:20: error: cannot convert ‘std::shared_ptr<JSBSim::FGInitialCondition>’ to ‘JSBSim::FGInitialCondition*’ in assignment
  335 |   fgic=fdmex->GetIC();
      |        ~~~~~~~~~~~~^~
      |                    |
      |                    std::shared_ptr<JSBSim::FGInitialCondition>

@agodemar Sir, I have been reading and following your papers for quite some time. It's commendable the contribution/efforts made by your lab for the development of JSBSim and appreciate the recent work you have published titled 
"A Deep Reinforcement Learning Control Approach for High-Performance Aircraft".

I request you to help me out with the FGTrimAnalysis code and mentioned error.

**Comment by agodemar** [comment]
@rishi1906 the `FGTrimAnalysis` related functionalities were written years ago, and you get errors like the one you posted above if you simply throw old code into the actual compilation workflow. This means some work is required to adapt that code to the new types we are using. If you have the possibility to do it, it shouldn't be so difficult, provided you know how to program in C++.
The compiler is so clear in the log you posted above:

In FGTrimAnalysis.cpp at line 335, there's a mismatch of types. you should modify `FGTrimAnalysis.cpp` by declaring `fgic` as `std::shared_ptrJSBSim`. Other errors will probably occur due to the same reason.

**Comment by rishi1906** [comment]
Thanks for the response. I anticipated the same, sure will fix the code according to the latest release of JSBSim, and yes I am comfortable with the programming environment.

**Reply by agodemar** [reply]
@rishi1906 that's great, please let me know about your progresses here in this thread.

Be advised that the trim analysis uses a variant of the Nelder-Mead algorithm that was published years ago by Virginia Torczon.
Nowadays, this algorithm is included into the fmincon and `fminsearch` `functions` (within the Optimization Toolbox).

**Comment by seanmcleod** [comment]
@rishi1906 are you planning on making use of JSBSim within a Matlab/Simulink environment via the S-Function interface, or are you planning on using JSBSim stand alone outside of a Matlab/Simulink environment?

**Comment by rishi1906** [comment]
Sir, I have started resolving the issues, I think I will be able to update them as mostly its conversion from string to SGPath and some updated Set and Get functions in addition to updating to shared_pointers as you rightly suggested. 

Regarding, fmincon I suppose that's part of MATLAB environment. I am a core C++ guy. 

@seanmcleod I am using JSBSim with Ardupilot SITL, so its outside MATLAB as I need to attach it with Software in the loop simulation and open source Optimizer routine.

**Comment by seanmcleod** [comment]
Thanks for the additional context. Out of interest is the current trim implementation in JSBSim not sufficient for what you need?

> Regarding, fmincon I suppose that's part of MATLAB environment. 

Yes, it's part of Matlab. I guess since you mentioned @agodemar's paper - "A Deep Reinforcement Learning Control Approach for High-Performance Aircraft" which made use of Matlab/Simulink he possibly assumed you were also planning on using Matlab/Simulink.

**Comment by rishi1906** [comment]
I debugged the current trim implementation in JSBSim, which uses a pairwise (state and control) method that does down the lane bisection method to update the guess, I am not fully sure why I am getting wdot not trimmable error for a custom aircraft I want to simulate in SITL. Work done by Professor @agodemar for general trim looks more promising for that matter. I am hoping it will resolve wdot no trimmable error.

The current trim solution in JSBSim, only solves for three states for example Ground Trim, but I guess all states should be trimmed according to the flight envelope as done @agodemar. I am surprised why current trim method has not been updated since 1999.

**Reply by agodemar** [reply]
@rishi1906 if you get good results by revamping the TrimAnalysis code, I hope you could pass the sources to us in order to have it back in the master branch. 

Are you working in a forked repo?

**Comment by rishi1906** [comment]
Sure, I will share the updated code. Yes, I have forked working with local files. I will need your help with the code you wrote.

**Comment by seanmcleod** [comment]
> I am surprised why current trim method has not been updated since 1999

That's why I was asking whether it wasn't sufficient for your needs, given it appears to have been for 90% of JSBSim users for the last 24 years :wink:

If you can share your custom aircraft I could take a look and see if I can see what the trim issue is.

**Comment by rishi1906** [comment]
@seanmcleod that would be great! Till I get correct stability derivatives for modified Talon (referred to as custom aircraft), could you check the attached XML file of X8 RC UAV. Please find attached initial condition file. I have taken this XML file from github just to test it with JSBSim. For full trim I get following output 

Full Trim

  Trim failed
  Trim Results: 
       Angle of Attack:   7.50  wdot:  3.07e+01 Tolerance: 1e-03  Failed
              Throttle:   0.00  udot: -1.43e+00 Tolerance: 1e-03  Failed
            Pitch Trim:   0.00  qdot:  7.79e-01 Tolerance: 1e-04  Failed
            Roll Angle:   0.00  vdot: -1.01e-04 Tolerance: 1e-03  Passed
              Ailerons:  -0.00  pdot:  0.00e+00 Tolerance: 1e-04  Passed
                Rudder:   0.00  rdot:  0.00e+00 Tolerance: 1e-04  Passed

[x8.zip](https://github.com/JSBSim-Team/jsbsim/files/13399298/x8.zip)

I tried to trim it (in MATLAB) for level flight and got these values

AoA     Elevator_deflection    V_inf     Thrust
    ____    ___________________    ______    ______

    -0.5           1.1232          52.573    13.807
       0          0.77393          35.453    6.1801
     0.5           0.4247           28.52    3.9987
       1         0.075472           24.52    3.0018
     1.5         -0.27376          21.837    2.4545
       2         -0.62299          19.878    2.1254
     2.5         -0.97222          18.366    1.9188
       3          -1.3215          17.154    1.7878
     3.5          -1.6707          16.154    1.7067
       4          -2.0199          15.311    1.6604
     4.5          -2.3691          14.587     1.639
       5          -2.7184          13.957    1.6365
     5.5          -3.0676          13.402    1.6482
       6          -3.4168          12.909    1.6712
     6.5          -3.7661          12.466    1.7032
       7          -4.1153          12.066    1.7424
     7.5          -4.4645          11.702    1.7876
       8          -4.8138          11.369    1.8378
     8.5           -5.163          11.063    1.8921
       9          -5.5122          10.781      1.95
     9.5          -5.8615          10.519    2.0109
      10          -6.2107          10.275    2.0744
Note : I have been working on JSBSim since last 1 month only, kindly ignore if I don't know some obvious things.

**Reply by rishi1906** [reply]
Correction,

 AoA     Elevator_deflection    V_inf     Thrust
    ____    ___________________    ______    ______

    -1.5            8.704          55.061    27.597
      -1           7.6942          36.515    12.118
    -0.5           6.6844          29.234     7.833
       0           5.6746          25.076      5.87
     0.5           4.6648          22.303    4.7732
       1           3.6549          20.284    4.0933
     1.5           2.6451          18.729    3.6462
       2           1.6353          17.485    3.3419
     2.5           0.6255           16.46    3.1319
       3         -0.38432          15.596    2.9871
     3.5          -1.3941          14.856    2.8895
       4           -2.404          14.211     2.827
     4.5          -3.4138          13.644    2.7914
       5          -4.4236           13.14    2.7768
     5.5          -5.4334          12.688     2.779
       6          -6.4432          12.279    2.7948
     6.5           -7.453          11.908    2.8217
       7          -8.4629          11.568    2.8579
     7.5          -9.4727          11.256    2.9019
       8          -10.482          10.968    2.9525
     8.5          -11.492          10.701    3.0088
       9          -12.502          10.453    3.0699
     9.5          -13.512          10.221    3.1353
      10          -14.522          10.003    3.2044

```xml
<?xml version="1.0"?>
<initialize name="basic_ic">
  <!--
    Basic set of initial conditions at 1000 ft with engines running.
  -->
  <ubody unit="FT/SEC">     42.9462 </ubody> 
  <vbody unit="FT/SEC">     0.0  </vbody>
  <wbody unit="FT/SEC">     3.7573  </wbody>
  <longitude unit="DEG">    0.0 </longitude> 
  <latitude unit="DEG">     0.0  </latitude>
  <phi unit="DEG">          0.0  </phi>  <!-- roll -->
  <theta unit="DEG">        5.0  </theta>  <!-- pitch -->
  <psi unit="DEG">          0.0  </psi>  <!-- yaw -->
  <altitude unit="FT">      1000.0 </altitude>
  <elevation unit="FT">     0.00000001 </elevation> <!-- bug in JSBSim makes 0 cause NaNs -->
  <winddir unit ="DEG">     90.0 </winddir> <!-- wind direction -->
  <vwind unit ="FT/SEC">    0.0 </vwind> <!-- wind speed -->
  <running>                  -1 </running> <!-- starts all engines -->
</initialize>
```

```
 Full Trim
  Sorry, qdot doesn't appear to be trimmable

  Trim failed
  Trim Results: 
       Angle of Attack:   4.81  wdot: -1.01e+00 Tolerance: 1e-03  Failed
              Throttle:   0.00  udot: -2.52e+00 Tolerance: 1e-03  Failed
            Pitch Trim:   0.00  qdot: -2.86e+00 Tolerance: 1e-04  Failed
            Roll Angle:   0.00  vdot: -8.20e-10 Tolerance: 1e-03  Passed
              Ailerons:   0.00  pdot:  1.29e-14 Tolerance: 1e-04  Passed
                Rudder:   0.00  rdot: -1.45e-14 Tolerance: 1e-04  Passed
```

**Comment by seanmcleod** [comment]
You didn't supply the engine file.

```sh
  Propulsion:
No filename given.


In file ../aircraft/x8/x8.xml: line 79
Could not open file: electric800W.xml

Aircraft propulsion element has problems in file Path "../aircraft/x8/x8.xml"
```

Is it identical to the existing electric engine files in JSBSim, just with the power set to 800W?

In your Matlab examples above, are all of the entries valid trim points? 

What is `V_inf`? And what units?

Also the units for the `Thrust` column?

**Comment by rishi1906** [comment]
Oh yes it's a simple JSBSim electric engine file.

```xml
<?xml version="1.0"?>

<electric_engine name="electric">
 <!-- This is a basic model of the electric motor used to power the x8
 -->
 <power unit="WATTS">  800.0 </power>
</electric_engine>

```

Yes, all are valid trim points.
V_inf is velocity in x direction (basically ubody in init file (XML), V_inf is in m/sec. 
Thrust is in Newton.
Attaching matlab code.
```matlab

% code to tabulate the equilibrium conditions of X8 aircraft

clear all; close all; clc

S = 0.75;           % Wing area in m2
m = 3.364;          % Aircraft weight in kg
rho = 1.225;        % Density of air in kg/m3
g = 9.81;           % Acceleration due to gravity in m/s2

% Aerodynamic coefficients

CL0 = 0.0867;
CLalp = 4.0203;
CLde = 0.2781;

CD0 = 0.0197;
CDalp1 = 0.079;
CDalp2 = 1.0555;
CDde = 0.0633;

Cm0 = 0.0227;
Cmalp = -0.4629;
Cmde = -0.2292;

% CD0 = 0.0107;
% CDalp1 = -0.00955;
% CDalp2 = 1.1;
% CDde = 0.0196;
% 
% CL0 = 0.0477;
% CLalp = 4.06;
% CLde = 0.7;
% 
% Cm0 = 0.00439;
% Cmalp = -0.227;
% Cmde = -0.325;

% Assume different values of alpha
AoA = -1.5:0.5:10;         % aoa in degrees
alpha = AoA*pi/180;     % aoa in radians

% zero moment
de = -(Cm0 + Cmalp*alpha)/Cmde;
Elevator_deflection = de*180/pi;        % in degrees

% Lift balance
Vsq = (m*g/(0.5*rho*S))./(CL0 + CLalp*alpha + CLde*de);
V_inf = sqrt(Vsq);

% Thrust measurement in Newtons
Thrust = (0.5*Vsq*rho*S).*(CD0 + CDalp1*alpha + CDalp2*(alpha.^2) + CDde*(de.^2));

AoA = AoA';
Elevator_deflection = Elevator_deflection';
V_inf = V_inf';
Thrust = Thrust';

X8_Equilibrium = table(AoA, Elevator_deflection, V_inf, Thrust)
```

**Comment by seanmcleod** [comment]
Ah, so you're not actually getting Matlab to come up with a trim solution, you've come up with a calculation to for a given AoA work out the elevator deflection needed to zero out the pitch moment due to $C_{m{_\alpha}}$ then the airspeed needed for the lift to equal the weight for the given alpha and finally the thrust needed for the drag for the given alpha.

You're assuming the MRP (Moment Reference Point) is coincident with the cg, which is the case for this model, although I see a lot of stuff is coincident at `(0, 0, 0)`, the cg, the MRP and the engine thruster.

One minor thing you haven't taken into account in your calculation is the thrust angle.

Currently JSBSim is saying `qdot doesn't appear to be trimmable`, i.e. it isn't able to zero out the pitch moment.

**Reply by rishi1906** [reply]
> Ah, so you're not actually getting Matlab to come up with a trim solution, you've come up with a calculation to for a given AoA work out the elevator deflection needed to zero out the pitch moment due to then the airspeed needed for the lift to equal the weight for the given alpha and finally the thrust needed for the drag for the given alpha.

Yes, you are right, it should work, isn't it? 

X8 has blended wing design with pusher configuration, shall I change location of the thruster to be at the end of fuselage?

> One minor thing you haven't taken into account in your calculation is the thrust angle.

I have taken those equations from the paper mentioned in the reference in x8.xml file. 
`
"Aerodynamic modeling of the Skywalker X8 Fixed-Wing Unmanned Aerial Vehicle"`

The thruster is oriented at 0,0,0 deg.

**Comment by seanmcleod** [comment]
> The thruster is oriented at 0,0,0 deg.

Yes it is, but if you're trimming for level flight gamma = 0 and say an AoA of 10 deg, then the pitch attitude, theta = 10 deg nd then the thrust line is oriented at 10 deg now relative to the ground. So some of the thrust vector is now providing a force in the lift direction, so you would need slightly less lift, but also slightly more thrust since the drag is parallel to the flight path etc. A trim solution needs to take this into account.

**Comment by seanmcleod** [comment]
One thing I noticed is that the aircraft seems to be very overpowered with an 800W electric engine, so trim wise it's trying to trim around throttle commands on the order of 0.00001.

So I dropped the power down to 4W.

So now you can see that the trim solution is looking at a throttle command of around 0.23. But still failing.

```sh
  Full Trim
  Sorry, qdot doesn't appear to be trimmable

  Trim failed
  Trim Results: 
       Angle of Attack:   2.97  wdot: -2.57e-01 Tolerance: 1e-03  Failed
              Throttle:   0.23  udot: -7.11e-02 Tolerance: 1e-03  Failed
            Pitch Trim:   0.00  qdot: -6.66e-05 Tolerance: 1e-04  Passed
            Roll Angle:   0.00  vdot: -3.79e-19 Tolerance: 1e-03  Passed
              Ailerons:   0.00  pdot:  0.00e+00 Tolerance: 1e-04  Passed
                Rudder:   0.00  rdot:  0.00e+00 Tolerance: 1e-04  Passed
```

I used the following python code to get a feel for what is happening, which is how I noticed how over-powered the aircraft is.

You'll see I've used your 3 deg AoA trim point to choose an airspeed and elevator angle, and then run it for 14 time steps to see how much deviation there is.

```python
import jsbsim

fdm = jsbsim.FGFDMExec('..\\') # The path supplied to FGFDMExec is the location of the folders "aircraft", "engines" and "systems"

fdm.load_model('x8') 

# Set engines running
fdm['propulsion/engine/set-running'] = 1

fdm['ic/h-sl-ft'] = 1
fdm['ic/vc-kts'] = 30.31882
fdm['ic/gamma-deg'] = 0
fdm['ic/alpha-deg'] = 3
fdm['ic/theta-deg'] = 3

fdm['fcs/throttle-cmd-norm'] = 0.225
fdm['fcs/elevator-cmd-norm'] = -0.012810666

fdm.run_ic() # Initialize the aircraft with initial conditions

#fdm['simulation/do_simple_trim'] = 1

elevator_pos = fdm['fcs/elevator-pos-rad']

cm0 = fdm['aero/coefficient/Cmo']
cm_alpha = fdm['aero/coefficient/Cmalpha']
cm_q = fdm['aero/coefficient/Cmq']
cm_de = fdm['aero/coefficient/CmDe']

alpha = fdm['aero/alpha-deg']

for i in range(14):
    alpha = fdm['aero/alpha-deg']
    gamma = fdm['flight-path/gamma-deg']
    cm0 = fdm['aero/coefficient/Cmo']
    cm_alpha = fdm['aero/coefficient/Cmalpha']
    cm_q = fdm['aero/coefficient/Cmq']
    cm_de = fdm['aero/coefficient/CmDe']
    
    M = fdm['moments/m-total-lbsft']
    aero_M = fdm['moments/m-aero-lbsft']
    propulsion_M = fdm['moments/m-prop-lbsft']
    fdm.run()

# Trim
try:
    fdm['simulation/do_simple_trim'] = 1

except RuntimeError as e:
    # The trim cannot succeed. Just make sure that the raised exception
    # is due to the trim failure otherwise rethrow.
    if e.args[0] != 'Trim Failed':
        raise
    
print('Done')
```

You can then also take a look at the `JSBoutX8.csv` file to look at a number of other states to see how they change over 0.1s.

You'll notice that there is a slight pitch up, a combination I think of slightly too much thrust giving a slight positive climb angle, which decreases AoA slightly, which means Cm-alpha decreases and so Cm_de results in a slight pitch rate etc.

**Comment by rishi1906** [comment]
@seanmcleod sorry for the late reply, yes what you mentioned is correct, it's indeed overpowered. 

I rechecked the configuration file and found some errors, CG locations, AERORP loc, locations of engine and thruster, and in controls as well some inputs were missing. JSBSim can do full trim successfully for updated configuration wrt the initial conditions provided as a solution from Matlab code.

 ```
Full Trim

  Trim successful
  Trim Results: 
       Angle of Attack:   0.48  wdot:  3.80e-04 Tolerance: 1e-03  Passed
              Throttle:   0.23  udot: -8.97e-05 Tolerance: 1e-03  Passed
            Pitch Trim:  -0.02  qdot:  6.39e-16 Tolerance: 1e-04  Passed
            Roll Angle:   0.00  vdot: -4.36e-17 Tolerance: 1e-03  Passed
              Ailerons:  -0.00  pdot:  1.58e-17 Tolerance: 1e-04  Passed
                Rudder:   0.00  rdot: -4.37e-13 Tolerance: 1e-04  Passed
```
Even upon getting trimmed, it doesn't maintain level flight in simulation (Flightgear), and altitude keeps dropping.
PFA updated XML and CSV files here. Could you check and let me know?

[x8.zip](https://github.com/JSBSim-Team/jsbsim/files/13413788/x8.zip)
[JSBout.csv](https://github.com/JSBSim-Team/jsbsim/files/13413794/JSBout.csv)

**Comment by seanmcleod** [comment]
Take a look at the last column `Direct Thrust (engine 0 in lbs)` in `JSBout.csv` it's 0.

What does your script/process look like in terms of performing the trim etc.?

It looks like the trim routine is figuring out how much thrust is required but then when you continue running the engine isn't running and/or the throttle has been reset to 0.

**Comment by rishi1906** [comment]
When I run the same script with standalone JSBSim code (without SITL) it gives a non-zero Direct Thrust value and maintains the height which indicates the trim is working correctly.

But same initial conditions with same configuration file when used in SITL behaves differently, what I has shared before. Any idea regarding Arduplane SITL ?

**Comment by seanmcleod** [comment]
Who is controlling the throttle during SITL? Log the throttle command and engine state during SITL.

**Reply by rishi1906** [reply]
That's what I need to figure out. Will update you once I get figure it out.

**Comment by seanmcleod** [comment]
Maybe you're getting caught out with ArduPilot's motor arming?

https://ardupilot.org/copter/docs/arming_the_motors.html

**Comment by rishi1906** [comment]
Yes, I figured it out today, it's due to arming only, as before arming Ardupilot doesn't give control command to JSBSim. It's working now (I had to tune the aircraft though using AutoTune mode in Ardupilot). 
I am facing a strange issue now. When I do full trim for X8 wrt following initial (trim) conditions for level flight, it works fine.

```xml
<?xml version="1.0"?>
<initialize name="basic_ic">
  <!--
    Basic set of initial conditions at 1000 ft with engines running.
  -->
  <ubody unit="M/SEC">     25.076 </ubody>
    <!-- <ubody unit="M/SEC">     50.00 </ubody> -->
  <vbody unit="FT/SEC">     0.0  </vbody>
  <wbody unit="FT/SEC">     0.0  </wbody>
  <!-- <longitude unit="DEG">    -122.372586 </longitude> 
  <latitude unit="DEG">     37.620397  </latitude> -->
  <longitude unit="DEG">    0.0 </longitude> 
  <latitude unit="DEG">     0.0  </latitude>
  <phi unit="DEG">          0.0  </phi>  <!-- roll -->
  <theta unit="DEG">        0.0  </theta>  <!-- pitch -->
  <psi unit="DEG">          0.0  </psi>  <!-- yaw -->
  <alpha unit="DEG">        0.0  </alpha>  <!-- alpha -->
  <altitude unit="FT">      1000.0 </altitude>
  <elevation unit="FT">     0.00000001 </elevation> <!-- bug in JSBSim makes 0 cause NaNs -->
  <winddir unit ="DEG">     90.0 </winddir> <!-- wind direction -->
  <vwind unit ="FT/SEC">    0.0 </vwind> <!-- wind speed -->
  <running>                  1 </running> <!-- starts all engines -->
</initialize>

```

But if I change lat long (say KSFO), trim fails. Trim shouldn't fail for different lat, and long, right?

**Reply by W-ShiYe** [reply]
Hello, I am also working on a skywalker x8 drone project recently, and I would like to ask you how you finally adjusted the relevant parameters to make the drone trim on the ground?

**Reply by rishi1906** [reply]
Hi @zzu-wangshuo what's the exact issue you are facing? It's been an year since I solved the issue, I will be able to help you better if you could share more detail.

**Comment by seanmcleod** [comment]
Welcome to the non-flat world :wink:

Graph what the trim routine computes for `fcs/aileron-cmd-norm` as you slowly increase the latitude from 0 deg to just over 7 deg when the trim starts failing.

The trim routine then fails with:

```sh
rdot doesn't appear to be trimmable
```

I haven't looked in detail, but what I'm guessing is that the increase in aileron deflection is generating a yaw moment which the trim routine can't cancel out with rudder sucessfully. So double-check the lateral and directional stability coefficients are 'correct'.

**Comment by seanmcleod** [comment]
Ah, I just took a quick glance at the X8 FDM, and I noticed that there is no yaw channel. So as soon as a yaw moment develops, e.g. due to aileron deflection, then the trim routine can't cancel it out with rudder, because there is no rudder/yaw control!

```xml
            <function name="aero/coefficient/Cnda">
                <description>Yaw moment due to aileron</description>
                <product>
                    <property>aero/qbar-area</property>
                    <property>metrics/bw-ft</property>
                    <property>fcs/effective-aileron-pos</property>
                    <value>0.1202</value>
                </product>
            </function>
```

**Reply by agodemar** [reply]
@seanmcleod good catch!

**Reply by rishi1906** [reply]
How does yaw moment generated due to aileron deflection depends on lat long?

How come with no rudder control for (0,0)lat,long, its able to trim rdot ?

X8 has blended wing config with no rudder control. What could be the way around to this problem ?

**Comment by seanmcleod** [comment]
> X8 has blended wing config with no rudder control. What could be the way around to this problem ?

Take a look at the B-2, it uses split elevons/flaperons for yaw control. 

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/c816db90-7c10-4856-94c2-e616fa54c701)

Or the B-21, comments from Rodney Robles on his LinkedIn post.

> I was delighted to see the maiden flight of the B-21 [hashtag#Rider](https://www.linkedin.com/feed/hashtag/?keywords=rider&highlightedUpdateUrns=urn%3Ali%3Aactivity%3A7129201002116698112) yesterday‼️
>
> In a photo posted by The Avionist, there was something that got my [hashtag#controlgeek](https://www.linkedin.com/feed/hashtag/?keywords=controlgeek&highlightedUpdateUrns=urn%3Ali%3Aactivity%3A7129201002116698112) attention, the two inboard flaperons deflected downwards and the outboard flaperon deflected upwards.
>
> With the B-21 control distribution, if you want a high lift configuration, you have to deflect the flaperons down, those which are in the inboard that can generate more lift, and to compensate for the pitch down induced by the previous, you have to deflect the outboard (the one having more pitch control power) up, having a total increase of lift, meanwhile maintaining a trimmed condition.
>
> However, there's a hidden good side effect here. This outer up, inner down flap distribution can generate [hashtag#differential](https://www.linkedin.com/feed/hashtag/?keywords=differential&highlightedUpdateUrns=urn%3Ali%3Aactivity%3A7129201002116698112) [hashtag#drag](https://www.linkedin.com/feed/hashtag/?keywords=drag&highlightedUpdateUrns=urn%3Ali%3Aactivity%3A7129201002116698112) and also can increase [hashtag#directional](https://www.linkedin.com/feed/hashtag/?keywords=directional&highlightedUpdateUrns=urn%3Ali%3Aactivity%3A7129201002116698112) [hashtag#static](https://www.linkedin.com/feed/hashtag/?keywords=static&highlightedUpdateUrns=urn%3Ali%3Aactivity%3A7129201002116698112) [hashtag#stability](https://www.linkedin.com/feed/hashtag/?keywords=stability&highlightedUpdateUrns=urn%3Ali%3Aactivity%3A7129201002116698112) (CnBeta>0), just like the B-2 did with the drag rudder (split flaps). 
>
> I love [hashtag#engineering](https://www.linkedin.com/feed/hashtag/?keywords=engineering&highlightedUpdateUrns=urn%3Ali%3Aactivity%3A7129201002116698112) ingenuity!

![image](https://github.com/JSBSim-Team/jsbsim/assets/824856/8cf0e55a-c893-4b22-8410-2678704402bf)

Now taking a quick look at the X-8 it looks like it only has two elevons, one on the port side and one on the starboard side. So it would be trickier to balance the pitch, roll and yaw moments required via differential movement.

**Reply by rishi1906** [reply]
Thanks for sharing, I agree using split elevons/flaperons is a good choice for yaw correction in blended wing design.

**Reply by seanmcleod** [reply]
Except that the actual X8 aircraft doesn't have split elevons/flaperons.

**Comment by rishi1906** [comment]
@seanmcleod could you guide me in deciding the parameters for the initialization file for the purpose of ground trim? 
JSBSim is not able to trim X8 using Ground Trim even if I give everything zero basically no movement at all (it is supposed to be just standing on the runway). 

```xml
<?xml version="1.0"?>
<initialize name="Start up at Canberra Model Aircraft Club">
  <latitude unit="DEG"> 0.0 </latitude>
  <longitude unit="DEG"> 0.0 </longitude>
  <altitude unit="M"> 0.0 </altitude>
  <vt unit="FT/SEC">      0.0 </vt>
  <gamma unit="DEG">      0.0 </gamma>
  <phi unit="DEG">        0.0 </phi>
  <theta unit="DEG">      0.0 </theta>
  <psi unit="DEG">        0.0 </psi>
</initialize>

```

It gives wdot not trimmable error


```
 Sorry, wdot doesn't appear to be trimmable
  Trim failed
  Trim Results: 
          Altitude AGL:   0.00  wdot:  3.21e+01 Tolerance: 1e-03  Failed
           Pitch Angle:   0.00  qdot:  2.92e-08 Tolerance: 1e-04  Passed
            Roll Angle:   0.00  pdot:  0.00e+00 Tolerance: 1e-04  Passed
```

Could it be due to AERORP and CG locations ?

**Reply by rishi1906** [reply]
Just resolved this, by giving a ground reaction to X8. but getting the following error while pre-arm check is performed in SITL.

```
AP: PreArm: AHRS: EKF3 Roll/Pitch inconsistent 41 deg. Wait or rebo
AP: PreArm: Check mag field (xy diff:260>100)
```

**Comment by seanmcleod** [comment]
FYI in terms of the trim routine in JSBSim.

https://github.com/JSBSim-Team/jsbsim/blob/5c31c25976fc1ea931103cce0c760cc49cea49fb/src/initialization/FGTrim.h#L74-L94

Now in terms of ground trim what is going to come into play are the ground reactions, typically between the gear and the ground. Now it looks like there is only as single ground reaction in the X8 at the moment, so maybe it can't balance, and/or the spring effect is too great etc.

```xml
    <ground_reactions>
      <contact type="STRUCTURE" name="Belly">
        <location unit="IN">
         <x> 0.0 </x>
         <y> 0.0 </y>
         <z> 0.0 </z>
        </location>
        <static_friction>  1.00 </static_friction>
        <dynamic_friction> 1.00 </dynamic_friction>
        <spring_coeff unit="LBS/FT">      10.00 </spring_coeff>
        <damping_coeff unit="LBS/FT/SEC"> 10.00 </damping_coeff>
      </contact>
    </ground_reactions>
```

**Reply by rishi1906** [reply]
Thanks for the detailed info, I updated Ground Reaction and Ground Trim is working now.

**Reply by shkarupa-dc-green** [reply]
@rishi1906, hi! Could you please share the updated version of the `x8.zip`?

**Comment by W-ShiYe** [comment]


**Reply by bcoconni** [reply]
@zzu-wangshuo could you please stop asking repeatedly to provide the x8 model ? This is annoying everyone who is watching the project.
Thanks.

**Comment by alanjt** [comment]
What has happened to JSBsim trim in recent years? It used to work, as long as the aircraft was actually trim-able. Now it fails most of the time. 
I have tested with JSBsim stand-alone scripts and with Flightgear. With turbojet aircraft at low airspeeds, the trim usually succeeds, but fails at higher speeds. Propeller (turbo-prop and piston engine) rarely trim successfully, even if the aircraft can be trimmed by piloted flying. In the case of propeller aircraft, the problem seems to be throttle/engine related, but I do not have time-histories of the trim algorithm to support this premise. 
In general, the example JSBsim aircraft in JSBsim´s git repository do not run as they have Flightgear dependencies, so I am not able to easily provide examples.
Regards
Alan

**Comment by seanmcleod70** [comment]
A couple of years ago I added some Python test scripts to the repo, see - https://github.com/JSBSim-Team/jsbsim/tree/master/examples/python, some of which perform a bunch of trimming.

In particular take a look at the Trim Envelope one - https://github.com/JSBSim-Team/jsbsim/blob/master/examples/python/TrimEnvelope.py

Using the 737 from the repo as a test case, the particular aircraft can easily be changed, it runs through hundreds of potential trim points, from 120KIAS to 460KIAS and for each speed flight path angles from -10 to +10. Generating a scatter plot for each successful trim point with colour coding to specify the amount of thrust/throttle required and the AoA required.

Here are the results from running it today, pretty sure it hasn't changed from 2 years ago.

![image](https://github.com/user-attachments/assets/3d21ee17-3067-46e5-be4f-226d8f67b828)

So yes, ideally you'd be able to supply an example aircraft FDM that can be run stand-alone with JSBSim so that we could use this script to see what it's trim envelope looks like, and then look into areas/examples where you would expect a successful trim point but there isn't one.

**Comment by seanmcleod70** [comment]
And an example for the C172p in the JSBSim repo.

![image](https://github.com/user-attachments/assets/3423eb47-2ba9-4bcd-a4e2-3304d6d2e009)
